[SampsonYe1999] iP

SampsonYe1999 opened the pull request:
# **Sam**

> "Hopefully one day we don't have to remember our schedules." - Sampson 

Sam can help you to remember tasks you need to complete. It's,

- text-based
- easy to learn 
- simple to use

All you need to do is,

1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you!

And its FREE!!

## Features:

- [ ] Managing tasks
- [ ] Managing deadlines 
- [ ] Managing event 
- [ ] Mark/unmark tasks 
- [ ] Find tasks 

Here is how it all starts in the Sam.java file:
```
    public static void main(String[] args) {
        new Sam().run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

SampsonYe1999 committed:
Level 0. Rename, Greet, Exit

SampsonYe1999 committed:
Level-0 updated

SampsonYe1999 committed:
Level 1. Echo

SampsonYe1999 committed:
Level 2. Add, List

SampsonYe1999 committed:
Level 3. Mark as Done.

SampsonYe1999 committed:
Level 4. ToDos, Events, Deadlines

SampsonYe1999 committed:
Level 5. Handle Errors

SampsonYe1999 committed:
Level 6. Delete

SampsonYe1999 committed:
Add Enum CommandType.java

There are multiple commands that Duke could recognize. Creating these
CommonTypes enable future updates on individual functions on a much
cleaner manner.

Add the function to read and save tasklist
New functions such as readTaskFromFile or loadTaskToFile have been
included so that Duke can store updated tasklists in the an existing
data file and read from it the next time its opened.

dillontkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+import java.io.*;
```
Should we avoid using the `*` form of import as per the [CS2103 Java Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)?

dillontkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+                System.out.println("Bye. Hope to see you again soon!");
+                break;
+            }
+            try{
```
Should there be whitespace after Java keywords and before `{`s? I noticed similar issues elsewhere.

dillontkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+                break;
+            default:
+                System.out.println("Skipped unknown task type in data file: " + taskType);
+                return null;
```
Are the `case` and `default` statements at the correct indentation level? I noticed the same issue in other places too.

dillontkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+                deleteTask(userInput, taskList);
+                break;
+            default:
+                throw new IllegalArgumentException("Hey, please choose from the following commands\n" +
+                        "if you want to add task, please use todo, deadline or event\n" +
+                        "if you want to mark or unmark task, please use mark or unmark\n" +
+                        "if you want delete a task, please use delete\n" +
+                        "if you want to view the existing task list, please enter list.");
```
Should like breaks be before or after operators like `+`?

dillontkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+            case "deadline":
+                String[] details = taskInfo.split("/by", 2);
+                if (details.length < 2) {
+                    throw new IllegalArgumentException("Invalid format for deadline, please provide a deadline using /by.");
```
Perhaps wrap lines so they do not exceed the [120 character limit](https://se-education.org/guides/conventions/java/intermediate.html#layout)?

dillontkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+                break;
+            }
+            try{
+                processCommand(userInput, taskList);
```
I like how you extracted the command processing to a separate function to make the `main` function more readable!

dillontkh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+        return task;
+    }
+    private static void displayList(List<Task> taskList) {
+        System.out.println("____________________________________________________________");
```
Maybe extract the printing of the bot messages to a method that automatically handles printing these lines?

dillontkh reviewed (commented):
Overall, I found your code quite readable. I noted a few coding standard violations, but nothing too serious. You may want to consider configuring your IDE to follow the coding standards, as well as using [checkstyle](https://se-education.org/guides/tutorials/checkstyle.html#using-checkstyle) to assist in adhering to those standards.

ChocoRagdoll commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+        }
+
+        String taskType = parts[0];
+        int isDone = Integer.parseInt(parts[1]);
```
Should we make this `int` a `boolean`?

ChocoRagdoll commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,255 @@
+                    throw new IllegalArgumentException("Invalid format for deadline, please provide a deadline using /by.");
+                }
+                String deadlineDescription = details[0].trim();
+                String deadline = details[1].trim();
```
I like how you use variable names to explain their usage.

ChocoRagdoll reviewed (commented):
LGTM. 
Some minor changes requested to follow the coding standard and code quality.

SampsonYe1999 committed:
Implement LocalDateTime class to deal with dates/times

The event and deadline classes are updated so that they can read
dates in given format instead of a string.

SampsonYe1999 committed:
Merge branch 'branch-Level-8'

SampsonYe1999 committed:
Add UI class and Storage class
This is to achieve better OOP design.

SampsonYe1999 committed:
Add TaskList class

Refactor TaskList into a class on its own for better OOP design

SampsonYe1999 committed:
Add Parser Class

Use the parser to separate the command from the task info when an
user input is given. This reduces the repeated string split method
in each of the command task. The parameters for each command are
updated.

SampsonYe1999 committed:
Merge branch 'branch-A-MoreOOP'

SampsonYe1999 committed:
Divide classes into packages

The classes are categorized into packages.

SampsonYe1999 committed:
Merge branch 'branch-A-Packages'

SampsonYe1999 committed:
Merge branch 'add-gradle-support'

SampsonYe1999 committed:
Add JUnit tests for Parser and ToDo classes.

SampsonYe1999 committed:
Change the output name of JAR file created

The name of the JAR file created has been changed to align with the
chatbox's name.

SampsonYe1999 committed:
Add JavaDoc Comments

Header comments are added to most most of the classes and methods.

SampsonYe1999 committed:
Tweak code to follow coding standard

Minor adjustments in layout/spacing to adhere to the convention.

SampsonYe1999 committed:
Add Find tasks features

A new command to find tasks with a specified keyword was added.

SampsonYe1999 committed:
Merge branch 'branch-A-CodingStandard'

SampsonYe1999 committed:
Merge branch 'branch-A-JavaDoc'

SampsonYe1999 committed:
Update javaDocs

Some javaDocs are updated and merge conflicts are resolved.

SampsonYe1999 committed:
Configure CheckStyle in and fix styles

Checkstyle plugins are enabled. Multiple style errors are fixed to
improve on readability of the code.

SampsonYe1999 committed:
Merge branch 'branch-A-Checkstyle'

SampsonYe1999 committed:
Fix more style errors

More style errors were detected and fixed.

SampsonYe1999 committed:
Improve parsing on Event Class

Certain problems detected from testing were fixed.

SampsonYe1999 committed:
Build GUI using JavaFX

A basic version of GUI has been created. Codes are improved and
certain code standard violations are adjusted.

SampsonYe1999 committed:
Merge branch 'branch-level-10'

SampsonYe1999 committed:
Add assertions in Parser

The parser splits the user input into a command type and task info.
Some commands do not accept task info but others do.

Assertions can help make explicit which commands require task info.

Let's use assertions to make sure the task info is only provided
when its needed.

SampsonYe1999 committed:
Fix codestyle errors in some files

Let's amend some string outputs since they are outdated.

SampsonYe1999 committed:
Fix more codestyle errors

Let's fix more codestyle errors that are found.

SampsonYe1999 committed:
Merge pull request #2 from SampsonYe1999/branch-A-CodeQuality

Fix more codestyle errors

SampsonYe1999 committed:
Add UpdateCommand

An update task command is added. This enables editing of the tasks'
information stored in the list.

SampsonYe1999 committed:
Merge branch 'branch-BCD-Extension'

SampsonYe1999 committed:
Implement better GUI

New pictures have been updated to better suit the style of the app.

SampsonYe1999 committed:
Add screenshot

A representative screenshot of the product has been uploaded.

SampsonYe1999 committed:
Rename UI.png to Ui.png

SampsonYe1999 committed:
Update ReadMe

The ReadMe doc is updated. Features and examples of usage are
listed. Certain changes for string output are made in some of the
classes for standardization.

SampsonYe1999 committed:
Merge branch 'master' of https://github.com/SampsonYe1999/ip

SampsonYe1999 committed:
Rectify branch error

This is just to highlight the branch-A-Assertions tab as it was not
done on a separate branch previously.

SampsonYe1999 committed:
Merge branch 'branch-A-Assertions'
